### PR TITLE
Make DtoH and HtoD transfers respect the current stream

### LINKED ISSUE #2702: The current CUDA stream is not honored by some operations
Hi, I think something is wrong with handling CUDA streams.  Could someone take a look?

(I'm using Python 2.7.12, CUDA 8.0, and PyTorch version 0.2.0.post1)

    #!/usr/bin/python
    # PyTorch CUDA stream test.

    import torch

    DIM = 2048
    M = torch.cuda.FloatTensor(DIM, DIM)

    def do_test(M, use_stream, fn):
        M.fill_(0.1)
        print 'Initial value = %.4f' % fn()

        # Create a new stream if requested.
        stream = torch.cuda.Stream() if use_stream else None
        with torch.cuda.stream(stream):
            for step in range(5):
                # Do random calculation.
                K = torch.matmul(M, M) / DIM
                M -= K
                print 'After step %d = %.4f' % (step, fn())

    fn = M.norm   # or 'M.sum' or 'lambda: M[0, 0]'

    print '=== Without using a new stream ==='
    do_test(M, False, fn)
    print '=== With using a new stream ==='
    do_test(M, True, fn)

Result on my machine (GTX 1080).  The second half keeps changing:

    === Without using a new stream ===
    Initial value = 204.8000
    After step 0 = 184.3197
    After step 1 = 167.7312
    After step 2 = 153.9943
    After step 3 = 142.4149
    After step 4 = 132.5118
    === With using a new stream ===
    Initial value = 204.8000
    After step 0 = 3.1469
    After step 1 = 3.1469
    After step 2 = 3.1469
    After step 3 = 3.1469
    After step 4 = 3.1469

Basically, when I'm using a different stream, computing any single float value returns random results.  Running nvprof shows that when the second do_test() runs, all the calculations happen in the new stream (13 below), but copying the result back to CPU happens in the default stream (7 below), and at the wrong moments before the correct value is ready.

    $ nvprof --print-gpu-trace ./cuda-stream-test.py

       Start  Duration  Size  Stream  Name
    ... (only showing the second half)
    2.28038s     320ns    4B       7  [CUDA memcpy DtoH]
    2.28086s  2.0956ms     -      13  sgemm_128x128x8_NN_vec [796]
    2.28132s     960ns    4B       7  [CUDA memcpy DtoH]
    2.28183s  1.1200us    4B       7  [CUDA memcpy DtoH]
    2.28194s     960ns    4B       7  [CUDA memcpy DtoH]
    2.28206s  1.0560us    4B       7  [CUDA memcpy DtoH]
    2.28218s  1.1200us    4B       7  [CUDA memcpy DtoH]
    2.28296s  147.40us     -      13  void kernelPointwiseApply2<TensorDivConstantOp<float>, float, float, unsigned int, int=-2, int=-2>(TensorInfo<TensorDivConstantOp<float>, float>, TensorInfo<float, float>, float, float) [814]
    2.28311s  212.52us     -      13  void kernelPointwiseApply2<TensorSubOp<float>, float, float, unsigned int, int=-2, int=-2>(TensorInfo<TensorSubOp<float>, float>, TensorInfo<float, float>, float, float) [827]
    2.28332s  93.890us     -      13  void kernelReduceAllPass1<TensorNormOp<float, int=2>, ReduceAdd<float, float>, ReduceAdd<float, float>, float, float, unsigned int, int=-2>(TensorInfo<float, ReduceAdd<float, float>>, ReduceAdd<float, float>, float, float, int=2, TensorNormOp<float, int=2>, float*) [857]
    2.28341s  2.8160us     -      13  void kernelReduceAllPass2<ReduceAdd<float, float>, float, unsigned int>(int, float, float, float*, float*) [866]
    2.28342s  2.0951ms     -      13  sgemm_128x128x8_NN_vec [892]
    ... (4 more calls to these kernels)

Please enlighten me if I'm doing something wrong.  Thanks!

### PR DESCRIPTION
**Pull request by zou3519**
### Summary
Fixes #2702. The bug is that cudaMemcpy calls occur on the default stream. This makes it so that those calls happen on the current stream.

### Test Plan

Run the code provided in #2702. Assert that junk data is not printed out.

### CODE REVIEW

**Comment by soumith** [comment]
do you need explicit `THCudaCheck(cudaStreamSynchronize(stream));` calls? I'd expect that if you make the cudaMemcpyAsync calls run in the `stream` stream, that would be enough to solve the bugs.

**Comment by zou3519** [comment]
I'm worried about the following two cases:
1) Modifying a CPU tensor while its data is being transferred to the GPU
```
y = x.cuda()  # x is some tensor on CPU
x.zero_()  # should wait for the HtoD transfer to finish
```

2) Trying to print out the result of a GPU operation
```
x = x.cuda()  # some tensor
some_tensor_op(x)
print(x)  # copies device to host, should wait for the some_tensor_op to finish
```

I'm not entirely sure these are problems (I'll try running them and see what happens)

**Comment by soumith** [comment]
that's a fair point. yes the synchronizes are needed then.

**Review by soumith — APPROVED** [review]
